File: frontend/src/types/optimization.ts
export interface Coordinate {
    x: number;
    y: number;
}

export interface GeoJSONPolygon {
    type: 'Polygon';
    coordinates: number[][][];
}

export interface GeoJSONLineString {
    type: 'LineString';
    coordinates: number[][];
}

export interface Building {
    id: number;
    building_type: 'academic' | 'dormitory' | 'social' | 'admin';
    shape_type: 'h_shape' | 'u_shape' | 'l_shape' | 'rectangular';
    floors: number;
    height: number;
    area: number;
    centroid: Coordinate;
    orientation: number;
    geometry: GeoJSONPolygon;
}

export interface Road {
    id: number;
    width: number;
    geometry: GeoJSONLineString;
}

export interface Driveway {
    building_id: number;
    geometry: GeoJSONLineString;
}

export interface Solution {
    solution_id: string;
    buildings: Building[];
    roads: Road[];
    driveways: Driveway[];
    objectives: Record<string, number>;
    constraints: Record<string, number>;
    rank: number;
    crowding_distance: number;
}

// Building type → color mapping
export const BUILDING_COLORS: Record<string, string> = {
    academic: '#4A90D9',   // Blue - formal/institutional
    dormitory: '#7CB342',  // Green - residential
    social: '#FFA726',     // Orange - vibrant/social
    admin: '#78909C'       // Gray - administrative
};

// Building type → 3D extrusion color (slightly darker for depth)
export const BUILDING_3D_COLORS: Record<string, string> = {
    academic: '#3A7BC8',
    dormitory: '#6BA32F',
    social: '#E89516',
    admin: '#607D8B'
};

File: backend/api/schemas/optimization.py
from pydantic import BaseModel
from typing import List, Dict, Any, Optional


class BuildingResponse(BaseModel):
    """Single building in optimization result."""
    id: int
    building_type: str          # "academic", "dormitory", "social", "admin"
    shape_type: str             # "h_shape", "u_shape", "l_shape", "rectangular"
    floors: int
    height: float               # meters (floors * 3.5)
    area: float                 # m² (polygon area)
    centroid: Dict[str, float]  # {"x": ..., "y": ...}
    orientation: float          # radians
    # GeoJSON Polygon coordinates for frontend
    geometry: Dict[str, Any]    # {"type": "Polygon", "coordinates": [...]}


class RoadResponse(BaseModel):
    """Single road segment."""
    id: int
    width: float
    # GeoJSON LineString
    geometry: Dict[str, Any]    # {"type": "LineString", "coordinates": [...]}


class DrivewayResponse(BaseModel):
    """Building-to-road connection."""
    building_id: int
    geometry: Dict[str, Any]    # {"type": "LineString", "coordinates": [...]}


class SolutionResponse(BaseModel):
    """Complete optimization solution."""
    solution_id: str
    buildings: List[BuildingResponse]
    roads: List[RoadResponse]
    driveways: List[DrivewayResponse]
    objectives: Dict[str, float]
    constraints: Dict[str, float]
    rank: int
    crowding_distance: float
